File: src/card/managed.ts
import type { LarkChannel } from '@larksuiteoapi/node-sdk';
import { log } from '../core/logger';

/**
 * Button-driven cards must be **CardKit 2.0 entities**, not raw interactive
 * JSON. `im.v1.message.patch` (what `channel.updateCard` does) silently no-ops
 * on a card sent as plain JSON via `channel.send({ card })` — the click just
 * flashes and reverts. So any card a user clicks to mutate in place goes
 * through here: create an entity → send a message that references it by
 * `card_id` → update the entity via `cardkit.v1.card.update` (monotonic
 * `sequence` so the server can't reorder/drop updates).
 */

interface ManagedEntry {
  cardId: string;
  sequence: number;
}

// Per-process; lost on restart, which is fine — a stale card just stops being
// updatable and the user re-triggers the flow to mint a fresh one.
const byMessageId = new Map<string, ManagedEntry>();

// The node-sdk dedups card-action callbacks for 12h keyed on
// `card:${messageId}:${openId}:${tag|name|option|value}` (SafetyPipeline /
// cardActionId), to swallow Feishu's re-deliveries of one click. But a console
// that updates ONE message in place reuses its messageId across every view, so
// a *second* click of any button whose value is unchanged (a cycle toggle, or
// 返回/设置 revisited) hashes to the same key and gets dropped — the click does
// nothing. We defeat that by stamping a fresh per-render token into every
// callback value: each (re)render gives its buttons new values, so a re-click
// after the card re-renders is a new key (fires), while a true re-delivery of
// the *same* rendered click keeps the same token (still deduped). Picked a
// `__r` key with sigils so it can't collide with real payload fields, and it
// stays well within cardActionId's 128-char value window.
let renderToken = 0;
function stampRenderToken(card: object): void {
  const token = (++renderToken).toString(36);
  const visit = (node: unknown): void => {
    if (Array.isArray(node)) {
      node.forEach(visit);
      return;
    }
    if (!node || typeof node !== 'object') return;
    const obj = node as Record<string, unknown>;
    const behaviors = obj.behaviors;
    if (Array.isArray(behaviors)) {
      for (const b of behaviors) {
        if (b && typeof b === 'object' && (b as { type?: unknown }).type === 'callback') {
          const v = (b as { value?: unknown }).value;
          if (v && typeof v === 'object') (v as Record<string, unknown>).__r = token;
        }
      }
    }
    for (const k of Object.keys(obj)) visit(obj[k]);
  };
  visit(card);
}

export interface ManagedCardSendResult {
  messageId: string;
  cardId: string;
}

/**
 * Create a CardKit entity and send a message referencing it. With `replyTo`
 * the card threads under the triggering message (im.v1.message.reply);
 * otherwise it posts top-level to `to`. Pass `replyInThread` when the
 * triggering message lives in a topic and the card should stay in it. Set
 * `receiveIdType: 'open_id'` (and pass an open_id as `to`) to DM a user
 * directly — Feishu opens/uses the p2p chat — used to send the bind card to the
 * admin who added the bot to a group. Ignored when `replyTo` is set.
 */
/**
 * A just-created CardKit entity occasionally hasn't propagated when the message
 * that references it is sent — Feishu 400s with 230099 / ErrCode 11310 "cardid
 * is invalid", so the card silently fails to appear (intermittent). Only this
 * transient is safe to retry: Feishu rejected the message outright, so nothing
 * was created (no duplicate on resend). Genuine errors and network losses (where
 * a message MAY have been created) are NOT matched, so they never duplicate.
 */
function isCardIdNotReady(err: unknown): boolean {
  const data = (err as { response?: { data?: { code?: number; msg?: string } } })?.response?.data;
  return data?.code === 230099 || /11310|cardid is invalid/i.test(data?.msg ?? '');
}

export async function sendManagedCard(
  channel: LarkChannel,
  to: string,
  card: object,
  replyTo?: string,
  replyInThread = false,
  receiveIdType: 'chat_id' | 'open_id' = 'chat_id',
): Promise<ManagedCardSendResult> {
  stampRenderToken(card);
  const data = JSON.stringify(card);

  // One attempt = create the entity + send the message that references it.
  const attempt = async (): Promise<ManagedCardSendResult> => {
    const created = await channel.rawClient.cardkit.v1.card.create({ data: { type: 'card_json', data } });
    const cardId = (created as { data?: { card_id?: string } }).data?.card_id;
    if (!cardId) {
      throw new Error(`cardkit.card.create returned no card_id: ${JSON.stringify(created).slice(0, 200)}`);
    }
    const content = JSON.stringify({ type: 'card', data: { card_id: cardId } });
    let messageId: string | undefined;
    if (replyTo) {
      const sent = await channel.rawClient.im.v1.message.reply({
        path: { message_id: replyTo },
        data: { msg_type: 'interactive', content, reply_in_thread: replyInThread },
      });
      messageId = (sent as { data?: { message_id?: string } }).data?.message_id;
    } else {
      const sent = await channel.rawClient.im.v1.message.create({
        params: { receive_id_type: receiveIdType },
        data: { receive_id: to, msg_type: 'interactive', content },
      });
      messageId = (sent as { data?: { message_id?: string } }).data?.message_id;
    }
    if (!messageId) {
      throw new Error('send card-by-reference returned no message_id');
    }
    byMessageId.set(messageId, { cardId, sequence: 0 });
    return { messageId, cardId };
  };

  // Re-create + resend with a short backoff while the entity is still
  // propagating; bail to the real error on anything else.
  for (let i = 0; ; i++) {
    try {
      return await attempt();
    } catch (err) {
      if (i >= 2 || !isCardIdNotReady(err)) throw err;
      log.fail('card', err, { phase: 'managed-send', attempt: i, retry: true });
      await new Promise((r) => setTimeout(r, 400 * (i + 1)));
    }
  }
}

/**
 * Replace the whole card of a managed entity, keyed by the messageId that
 * carries it. Returns false (and logs) if we have no mapping — caller can fall
 * back to a fresh card. Sequence auto-increments per card.
 */
export async function updateManagedCard(
  channel: LarkChannel,
  messageId: string,
  card: object,
): Promise<boolean> {
  const entry = byMessageId.get(messageId);
  if (!entry) {
    // The silent failure that made "返回菜单又没用了" invisible: no mapping (e.g.
    // the card was sent before a restart — byMessageId is per-process). Log it
    // so the diagnosis isn't a black box; caller may fall back to a fresh card.
    log.info('card', 'managed-update-no-entry', { messageId, known: byMessageId.size });
    return false;
  }
  stampRenderToken(card);
  const data = JSON.stringify(card);
  const push = async (): Promise<void> => {
    entry.sequence += 1;
    await channel.rawClient.cardkit.v1.card.update({
      path: { card_id: entry.cardId },
      data: { card: { type: 'card_json', data }, sequence: entry.sequence, uuid: `u_${entry.cardId}_${entry.sequence}` },
    });
  };
  try {
    await push();
    return true;
  } catch (err) {
    // err 200810: the card is still in a previous click's interaction window.
    // Wait out the 3s callback window and retry once with the next sequence.
    log.fail('card', err, { phase: 'managed-update', cardId: entry.cardId, seq: entry.sequence, retry: true });
    await new Promise((r) => setTimeout(r, 3200));
    try {
      await push();
      return true;
    } catch (err2) {
      log.fail('card', err2, { phase: 'managed-update-retry', cardId: entry.cardId, seq: entry.sequence });
      return false;
    }
  }
}

/** True iff we hold the card_id mapping for this messageId. */
export function isManaged(messageId: string): boolean {
  return byMessageId.has(messageId);
}

/** Drop the mapping (card recalled / flow ended). */
export function forgetManagedCard(messageId: string): void {
  byMessageId.delete(messageId);
}

File: internal/card/managed.go
package card

// managed.go —— CardKit 实体托管（对齐 TS card/managed）。
// 按钮卡必须走 cardkit 实体（plain JSON 的 im.message.patch 对按钮静默失效）。
// sendManagedCard: create 实体 + 发引用消息 + stampRenderToken 反 12h cardActionId 去重。
// updateManagedCard: byMessageID→cardId 映射 + 单调 seq + 200810 点击窗口重试。

import (
	"context"
	"encoding/json"
	"fmt"
	"sync"
	"sync/atomic"
)

// ManagedRegistry CardKit 实体托管（进程内 messageId→cardId 映射）。
type ManagedRegistry struct {
	client      CardKitClient
	sleep       func(ms int64)
	mu          sync.Mutex
	byMessageID map[string]*managedEntry
	renderToken atomic.Int64
}

type managedEntry struct {
	cardID   string
	sequence int
}

// NewManagedRegistry 构造。
func NewManagedRegistry(client CardKitClient, sleep func(int64)) *ManagedRegistry {
	if sleep == nil {
		sleep = func(int64) {}
	}
	return &ManagedRegistry{client: client, sleep: sleep, byMessageID: map[string]*managedEntry{}}
}

// ManagedCardSendResult 发送结果。
type ManagedCardSendResult struct {
	MessageID string
	CardID    string
}

// SendManagedCard create 实体 + 发引用消息。230099 重试（最多 2 次）。
// replyTo 非空 → 回复（replyInThread 控话题内）；否则顶层发到 to（chat_id）。
func (r *ManagedRegistry) SendManagedCard(ctx context.Context, to string, card CardObject, replyTo string, replyInThread bool) (*ManagedCardSendResult, error) {
	r.stampRenderToken(card)
	data, _ := json.Marshal(card)
	for attempt := 0; ; attempt++ {
		cardID, err := r.client.CardCreate(string(data))
		if err != nil {
			if attempt >= 2 || !IsCardIdNotReady(err) {
				return nil, fmt.Errorf("managed.card.create: %w", err)
			}
			r.sleep(int64(400 * (attempt + 1)))
			continue
		}
		var msgID string
		if replyTo != "" {
			msgID, err = r.client.MessageReplyWithCard(replyTo, cardID, replyInThread)
		} else {
			msgID, err = r.client.MessageCreateWithCard(to, cardID)
		}
		if err != nil {
			if attempt >= 2 || !IsCardIdNotReady(err) {
				return nil, fmt.Errorf("managed.card.send: %w", err)
			}
			r.sleep(int64(400 * (attempt + 1)))
			continue
		}
		r.mu.Lock()
		r.byMessageID[msgID] = &managedEntry{cardID: cardID, sequence: 0}
		r.mu.Unlock()
		return &ManagedCardSendResult{MessageID: msgID, CardID: cardID}, nil
	}
}

// UpdateManagedCard 按 messageId 更新实体（单调 seq + 200810 重试）。无映射返回 false。
func (r *ManagedRegistry) UpdateManagedCard(ctx context.Context, messageID string, card CardObject) bool {
	r.mu.Lock()
	entry, ok := r.byMessageID[messageID]
	r.mu.Unlock()
	if !ok {
		return false
	}
	r.stampRenderToken(card)
	data, _ := json.Marshal(card)

	push := func() error {
		entry.sequence++
		uuid := fmt.Sprintf("u_%s_%d", entry.cardID, entry.sequence)
		return r.client.CardUpdate(entry.cardID, string(data), entry.sequence, uuid)
	}
	if err := push(); err == nil {
		return true
	}
	// 200810 点击窗口：等 3.2s 重试一次（下一个 seq）。
	r.sleep(3200)
	if err := push(); err == nil {
		return true
	}
	return false
}

// IsManaged 是否持有该 messageId 的 cardId 映射。
func (r *ManagedRegistry) IsManaged(messageID string) bool {
	r.mu.Lock()
	defer r.mu.Unlock()
	_, ok := r.byMessageID[messageID]
	return ok
}

// Forget 丢弃映射（卡片撤回/流程结束）。
func (r *ManagedRegistry) Forget(messageID string) {
	r.mu.Lock()
	defer r.mu.Unlock()
	delete(r.byMessageID, messageID)
}

// stampRenderToken 递归给每个 callback behavior 的 value 盖自增 __r token。
// 反飞书 SDK 12h cardActionId 去重（同卡重渲后再次点击是新 key）。
func (r *ManagedRegistry) stampRenderToken(card CardObject) {
	token := r.renderToken.Add(1)
	tokenStr := encodeBase36(token)
	visitManaged(card, tokenStr)
}

func visitManaged(node any, token string) {
	switch v := node.(type) {
	case map[string]any:
		stampCallbackBehaviors(v, token)
		for _, child := range v {
			visitManaged(child, token)
		}
	case []any:
		for _, child := range v {
			visitManaged(child, token)
		}
	case []map[string]any: // []CardElement（代码构造的 behaviors/elements）
		for _, child := range v {
			visitManaged(child, token)
		}
	}
}

// stampCallbackBehaviors 给 map 里 callback behaviors 的 value 盖 __r。
// behaviors 可能是 []any（JSON 解码）或 []map[string]any（CardElement 构造）。
func stampCallbackBehaviors(m map[string]any, token string) {
	var behaviors []map[string]any
	switch bv := m["behaviors"].(type) {
	case []any:
		for _, b := range bv {
			if bm, ok := b.(map[string]any); ok {
				behaviors = append(behaviors, bm)
			}
		}
	case []map[string]any:
		behaviors = bv
	}
	for _, bm := range behaviors {
		if bm["type"] == "callback" {
			if val, ok := bm["value"].(map[string]any); ok {
				val["__r"] = token
			}
		}
	}
}

func encodeBase36(n int64) string {
	if n == 0 {
		return "0"
	}
	const digits = "0123456789abcdefghijklmnopqrstuvwxyz"
	var buf [16]byte
	p := len(buf)
	for n > 0 {
		p--
		buf[p] = digits[n%36]
		n /= 36
	}
	return string(buf[p:])
}
